### PR TITLE
Client - Update Milestones and Message Node Events APIs

### LINKED ISSUE #404: Client - Update/Add Milestones, Message and Receipt Node Events API
[EVENT API DOC](https://studio.asyncapi.com/?url=https://raw.githubusercontent.com/iotaledger/tips/stardust-event-api/tips/TIP-0028/event-api.yml)
- [ ] milestone-info/latest - change topic string and update response
- [ ] milestone-info/confirmed - change topic and update response
- [ ] milestones - add new topic
- [ ] messages/transaction/tagged-data - add new topic
- [ ] message-metadata/referenced - change topic string
- [ ] message-metadata/{messageId} - change topic string
~~- [ ] receipts - add new topic~~

### PR DESCRIPTION
**Pull request by rebinmannady93**
# Description of change

Fixes #404 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

get_event_message.c
test_milestones_events.c
test_messages_events.c

## Change checklist

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/client/api/events/sub_messages_metadata.c [hunk 1]
```diff
@@ -12,10 +12,10 @@ int event_subscribe_msg_metadata(event_client_handle_t client, int *mid, char co
     printf("[%s:%d]: Message Id length is invalid\n", __func__, __LINE__);
     return -1;
   }
-  // Buffer to store topic string : messages/{messageid}/metadata
-  char topic_buff[BIN_TO_HEX_BYTES(IOTA_MESSAGE_ID_BYTES) + 21] = {0};
+  // Buffer to store topic string : messages-metadata/{messageid}
+  char topic_buff[BIN_TO_HEX_BYTES(IOTA_MESSAGE_ID_BYTES) + 20] = {0};
```
Comment (oopsmonk) [inline src/client/api/events/sub_messages_metadata.c]: prefix `0x` is missing
```suggestion
  // Buffer to store topic string : messages-metadata/0x{messageid}
  char topic_buff[BIN_TO_HEX_BYTES(IOTA_MESSAGE_ID_BYTES) + 21] = {0};
```

File: examples/client/get_event_messages.c [hunk 2]
```diff
@@ -184,7 +198,7 @@ void process_event_data(event_client_event_t *event) {
            strstr(event->topic, "messages/tagged-data/") != NULL) {
     print_serialized_data(event->data, event->data_len);
   }
-  // check for topic messages/{messageId}/metadata
+  // check for topic message-metadata/{messageId}
   else if ((strstr(event->topic, "messages/") != NULL) && (strstr(event->topic, "/metadata") != NULL)) {
```
Comment (oopsmonk) [inline examples/client/get_event_messages.c]: if condition needs an update as well.
```suggestion
  // check for topic message-metadata/{messageId}
  else if (strcmp(event->topic, "messages-metadata/") != NULL ) {
```

**Review by oopsmonk — CHANGES_REQUESTED** [review]
tag parameters need `0x` prefix in the topic.

```
// in event_sub_tx_msg_tagged_data()
sprintf(topic_buff, "messages/transaction/tagged-data/0x%s", tag);
// in event_sub_msg_tagged_data()
sprintf(topic_buff, "messages/tagged-data/0x%s", tag);
```

File: src/client/api/events/sub_messages_metadata.c (hunk 1 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_messages_metadata.c]: 0x was considered in topic as below, but was not explicitely mentioned in the comment. We need only len 20 for messages-metadata/0x. Added 0x in comment also. Thanks.
sprintf(topic_buff, "message-metadata/0x%s", msg_id);

File: examples/client/get_event_messages.c (hunk 2 above)
Comment (rebinmannady93) [reply examples/client/get_event_messages.c]: I missed that. Thank You.

**Comment by rebinmannady93** [comment]
> tag parameters need `0x` prefix in the topic.
> 
> ```
> // in event_sub_tx_msg_tagged_data()
> sprintf(topic_buff, "messages/transaction/tagged-data/0x%s", tag);
> // in event_sub_msg_tagged_data()
> sprintf(topic_buff, "messages/tagged-data/0x%s", tag);
> ```

Done.

File: examples/client/node_info.c [hunk 3]
```diff
@@ -14,6 +14,7 @@ int main(void) {
   iota_client_conf_t ctx = {.host = "localhost", .port = 443, .use_tls = true};
 
   res_node_info_t *info = res_node_info_new();
+
```
Comment (oopsmonk) [inline examples/client/node_info.c]: should keep this file untouched

File: src/client/api/events/sub_messages_metadata.c [hunk 4]
```diff
@@ -12,10 +12,10 @@ int event_subscribe_msg_metadata(event_client_handle_t client, int *mid, char co
     printf("[%s:%d]: Message Id length is invalid\n", __func__, __LINE__);
     return -1;
   }
-  // Buffer to store topic string : messages/{messageid}/metadata
-  char topic_buff[BIN_TO_HEX_BYTES(IOTA_MESSAGE_ID_BYTES) + 21] = {0};
+  // Buffer to store topic string : messages-metadata/0x{messageid}
+  char topic_buff[BIN_TO_HEX_BYTES(IOTA_MESSAGE_ID_BYTES) + 20] = {0};
```
Comment (oopsmonk) [inline src/client/api/events/sub_messages_metadata.c:16]: the length of "messages-metadata/0x" string is 21.

File: src/client/api/events/sub_serialized_output.c [hunk 5]
```diff
@@ -26,10 +26,10 @@ int event_sub_tx_msg_tagged_data(event_client_handle_t client, int *mid, char co
     printf("[%s:%d]: Tag length is invalid\n", __func__, __LINE__);
     return -1;
   }
-  // 98 is the max length for string messages/transaction/tagged-data/{tag}, tag max len is 64 bytes
-  char topic_buff[98] = {0};
+  // 100 is the max length for string messages/transaction/tagged-data/0x{tag}, tag max len is 64 bytes
+  char topic_buff[100] = {0};
```
Comment (oopsmonk) [inline src/client/api/events/sub_serialized_output.c]: the maximum length is 64 bytes and the max hex-encoded-tag is 128, `BIN_TO_HEX_BYTES(TAGGED_DATA_TAG_MAX_LENGTH_BYTES)` , therefore buffer size should be 
```suggestion
  // 100 is the max length for string messages/transaction/tagged-data/0x{tag}, tag max len is 64 bytes
  char topic_buff[36+BIN_TO_HEX_BYTES(TAGGED_DATA_TAG_MAX_LENGTH_BYTES)] = {0};
```

File: src/client/api/events/sub_serialized_output.c [hunk 6]
```diff
@@ -39,10 +39,10 @@ int event_sub_msg_tagged_data(event_client_handle_t client, int *mid, char const
     printf("[%s:%d]: Tag length is invalid\n", __func__, __LINE__);
     return -1;
   }
-  // 86 is the max length for string messages/tagged-data/{tag}, tag max len is 64 bytes
-  char topic_buff[86] = {0};
+  // 88 is the max length for string messages/tagged-data/0x{tag}, tag max len is 64 bytes
+  char topic_buff[88] = {0};
```
Comment (oopsmonk) [inline src/client/api/events/sub_serialized_output.c]: the same as above.

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/client/api/events/sub_messages_metadata.c (hunk 4 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_messages_metadata.c:16]: Sorry, topic had a change. It is "message-metadata/0x", 's' was removed from message's' ie why length is 20. I have updated the comment.

File: src/client/api/events/sub_serialized_output.c (hunk 5 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_serialized_output.c]: Topic buffer length is changed to `36+BIN_TO_HEX_BYTES(TAGGED_DATA_TAG_MAX_LENGTH_BYTES)`. Tag hex string len shoud be 128 not 64.

To keep the tagged data input similar to the restful method, now accepting `tag` in bytes and `tag_len` as a sperate parameter and tag is converted to hex encoded string.

File: src/client/api/events/sub_serialized_output.c [hunk 7]
```diff
@@ -13,36 +13,50 @@ int event_sub_txn_included_msg(event_client_handle_t client, int *mid, char cons
     printf("[%s:%d]: Transaction id length is invalid\n", __func__, __LINE__);
     return -1;
   }
-  // 95 is the max length for string transactions/{transactionId}/included-message
+  // 97 is the max length for string transactions/{transactionId}/included-message
   char topic_buff[97] = {0};
```
Comment (matjazv) [inline src/client/api/events/sub_serialized_output.c]:
```suggestion
  // 97 is the max length for string transactions/0x{transactionId}/included-message
  char topic_buff[33 + BIN_TO_HEX_BYTES(IOTA_TRANSACTION_ID_BYTES)] = {0};
```

**Review by matjazv — CHANGES_REQUESTED** [review]

File: src/client/api/events/sub_serialized_output.c (hunk 7 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_serialized_output.c]: Done.

**Review by matjazv — APPROVED** [review]

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**